File: generators/eigen.rb
#!/usr/bin/env ruby
# General Eigen value problem generator
# Usage:
#   ruby generators/eigen.rb <output_file> [n] [density_A] [density_B] [--verbose]
# Produces two symmetric sparse matrices A and B in triplet form:
#   row,col,value per line (0-based indices)

require 'matrix'

n = argv[0] || 100
density_A = argv[0]  || 0.01 
density_B = argv[0]  || 0.005

def generate_symmetric_sparse_matrix(n, density, scale_factor = 1.0)

  triplets = []
  
  (0...n).each do |i|
    diag_value = (i + 1).to_f * scale_factor + rand * 5.0
    triplets << [i, i, diag_value]
  end
  
  num_nonzeros = (n * n * density / 2).to_i  
  
  num_nonzeros.times do
    i = rand(n)
    j = rand(n)
    next if i == j 
    next if i > j 
    
    value = (rand - 0.5) * 2.0 * scale_factor
    triplets << [i, j, value]
    triplets << [j, i, value]  
  end
  
  hash = {}
  triplets.each do |i, j, v|
    key = [i, j]
    hash[key] ||= 0.0
    hash[key] += v
  end

  result = []
  hash.each do |(i, j), v|
    result << [i, j, v] if v.abs > 1e-10
  end
  
  result.sort_by { |i, j, v| [i, j] }
end

def save_triplets(filename, triplets)
  File.open(filename, 'w') do |f|
    triplets.each do |i, j, v|
      f.puts "#{i} #{j} #{v}"
    end
  end
end

triplets_A = generate_symmetric_sparse_matrix(n, density_A, 1.0)
triplets_B = generate_symmetric_sparse_matrix(n, density_B, 0.5)

# B is reaaly good in this case, strongly diagonally dominant
triplets_B.each do |triplet|
  if triplet[0] == triplet[1]
    triplet[2] += 10.0
  end
end

# Сохраняем данные
puts
puts "Saving matrices..."
save_triplets("../data/matrix_A_1000.txt", triplets_A)
save_triplets("../data/matrix_B_1000.txt", triplets_B)

puts
puts "=" * 60
puts "Matrix statistics:"
puts "=" * 60
puts "Matrix A:"
puts "  Size: #{n} x #{n}"
puts "  Nonzeros: #{triplets_A.size}"
puts "  Density: #{(triplets_A.size.to_f / (n * n) * 100).round(3)}%"
puts "  Estimated memory: #{(triplets_A.size * 12 / 1024.0).round(2)} KB"
puts
puts "Matrix B:"
puts "  Size: #{n} x #{n}"
puts "  Nonzeros: #{triplets_B.size}"
puts "  Density: #{(triplets_B.size.to_f / (n * n) * 100).round(3)}%"
puts "  Estimated memory: #{(triplets_B.size * 12 / 1024.0).round(2)} KB"
puts

File: python/src/bsmp/generators/eigen.py
#!/usr/bin/env python
import random
from os.path import dirname, exists


# Generate a symmetric sparse matrix as a list of [row, col, value] triplets
def generate_symmetric_sparse_matrix(
    n: int, density: float, scale_factor: float = 1.0
) -> list[list]:
    triplets = []

    for i in range(n):
        diag_value = (i + 1) * scale_factor + random.random() * 5.0
        triplets.append([i, i, diag_value])

    num_nonzeros = int(n * n * density / 2)

    for _ in range(num_nonzeros):
        i = random.randrange(n)
        j = random.randrange(n)
        if i == j:
            continue
        if i > j:
            continue

        value = (random.random() - 0.5) * 2.0 * scale_factor
        triplets.append([i, j, value])
        triplets.append([j, i, value])

    hash_map = {}
    for i, j, v in triplets:
        key = (i, j)
        hash_map[key] = hash_map.get(key, 0.0) + v

    result = []
    for (i, j), v in hash_map.items():
        if abs(v) > 1e-10:
            result.append([i, j, v])

    result.sort(key=lambda t: (t[0], t[1]))
    return result


# Write triplets to a file, one "row col value" per line
def save_triplets(filename: str, triplets: list[list]) -> None:
    with open(filename, "w") as f:
        for i, j, v in triplets:
            f.write(f"{i} {j} {v}\n")


# Generate the A and B matrices of a generalized eigenvalue problem
def generate_eigen_matrices(
    output_filename_A: str,
    output_filename_B: str,
    n: int,
    density_A: float,
    density_B: float,
) -> None:
    if n <= 0:
        raise ValueError(f"n must be positive, got {n}")
    if density_A <= 0 or density_B <= 0:
        raise ValueError(
            f"densities must be positive, got {density_A}, {density_B}"
        )
    if not output_filename_A or not output_filename_B:
        raise ValueError("output filename is empty")
    if not exists(dirname(output_filename_A)):
        raise ValueError(
            f"output directory does not exist: {dirname(output_filename_A)}"
        )
    if not exists(dirname(output_filename_B)):
        raise ValueError(
            f"output directory does not exist: {dirname(output_filename_B)}"
        )

    triplets_A = generate_symmetric_sparse_matrix(n, density_A, 1.0)
    triplets_B = generate_symmetric_sparse_matrix(n, density_B, 0.5)

    # B is really good in this case, strongly diagonally dominant
    for triplet in triplets_B:
        if triplet[0] == triplet[1]:
            triplet[2] += 10.0

    print()
    print("Saving matrices...")
    save_triplets(output_filename_A, triplets_A)
    save_triplets(output_filename_B, triplets_B)

    print()
    print("=" * 60)
    print("Matrix statistics:")
    print("=" * 60)
    print("Matrix A:")
    print(f"  Size: {n} x {n}")
    print(f"  Nonzeros: {len(triplets_A)}")
    print(f"  Density: {round(len(triplets_A) / (n * n) * 100, 3)}%")
    print(f"  Estimated memory: {round(len(triplets_A) * 12 / 1024.0, 2)} KB")
    print()
    print("Matrix B:")
    print(f"  Size: {n} x {n}")
    print(f"  Nonzeros: {len(triplets_B)}")
    print(f"  Density: {round(len(triplets_B) / (n * n) * 100, 3)}%")
    print(f"  Estimated memory: {round(len(triplets_B) * 12 / 1024.0, 2)} KB")
    print()
